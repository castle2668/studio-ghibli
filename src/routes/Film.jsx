import PropTypes from "prop-types";
import React from "react";
import { useFetcher, useLoaderData } from "react-router-dom";

import { getFilm, updateFilm } from "@/utils/films";

export async function loader({ params }) {
  const film = await getFilm(params.filmId);
  // Throw a 404 response in the loader
  if (!film) {
    throw new Response("", {
      status: 404,
      statusText: "Not Found",
    });
  }
  return film;
}

export async function action({ request, params }) {
  const formData = await request.formData();
  return updateFilm(params.filmId, {
    favorite: formData.get("favorite") === "true" ? "true" : "false",
  });
}

const Favorite = ({ film }) => {
  const fetcher = useFetcher();

  let { favorite } = film;
  // Optimistic UI:
  // If the fetcher has any formData being submitted, the star change to the new state immediately.
  // When the action is done, we're back to using the actual data.
  if (fetcher.formData) {
    favorite = fetcher.formData.get("favorite") === "true" ? "true" : "false";
  }

  return (
    <fetcher.Form method="post">
      <button
        type="submit"
        name="favorite"
        value={favorite === "true" ? "false" : "true"}
        aria-label={
          favorite === "true" ? "Remove from favorites" : "Add to favorites"
        }
      >
        {favorite === "true" ? "★" : "☆"}
      </button>
    </fetcher.Form>
  );
};

const Film = () => {
  const film = useLoaderData();

  return (
    <div id="film">
      <div>
        <img src={film.image || null} alt="" />
      </div>

      <div>
        <h1>
          {film.title}
          <Favorite film={film} />
        </h1>

        {film.original_title && (
          <p>
            <a
              target="_blank"
              href={`https://twitter.com/${film.original_title}`}
              rel="noreferrer"
            >
              {film.original_title}
            </a>
          </p>
        )}

        {film.description && <p>{film.description}</p>}
      </div>
    </div>
  );
};

Favorite.propTypes = {
  film: PropTypes.shape({
    favorite: PropTypes.string,
  }),
};
Favorite.defaultProps = {
  film: {},
};

export default Film;
