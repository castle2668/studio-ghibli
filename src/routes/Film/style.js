import styled from "styled-components";

export const Row = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;

export const Column = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

export const FilmContainer = styled.div`
  max-width: 64rem;

  & a[href*="twitter"] {
    display: flex;
    font-size: 2.4rem;
    color: #3992ff;
    text-decoration: none;
  }
  & a[href*="twitter"]:hover {
    text-decoration: underline;
  }
`;

export const FilmBanner = styled.img`
  display: block;
  width: 100%;
  border-radius: 2.4rem;
  object-fit: cover;
  box-shadow: 0px 0px 0.3rem 0.5rem #f2e1f2;
`;

export const FavoriteForm = styled.div`
  & form {
    display: flex;
    align-items: center;
    margin-top: 0.4rem;
    & button {
      box-shadow: none;
      font-size: 2.4rem;
      font-weight: 400;
      padding: 0;
      cursor: pointer;
    }
    & button[value="true"] {
      color: #a4a4a4;
    }
    & button[value="true"]:hover,
    & button[value="false"] {
      color: #eeb004;
    }
  }
`;

export const Title = styled.h1`
  display: flex;
  align-items: flex-start;
  gap: 1.6rem;
  text-align: center;

  font-size: 3.2rem;
  font-weight: 700;
  line-height: 1.2;
  & + p {
    margin: 0;
  }
  & + p + p {
    white-space: break-spaces;
  }
  &:focus {
    outline: none;
    color: hsl(224, 98%, 58%);
  }
  & ~ div {
    display: flex;
    gap: 0.8rem;
    margin: 1.6rem 0;
  }
`;

export const Score = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background: orange;
  color: #fff;
  width: 2.4rem;
  height: 2.4rem;
  box-shadow: 0px 0px 0.3rem 0.5rem orange;
`;

export const Description = styled.p`
  line-height: 2;
  text-align: justify;
`;
