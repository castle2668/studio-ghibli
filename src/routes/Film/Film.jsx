import PropTypes from "prop-types";
import React from "react";
import { useFetcher, useLoaderData } from "react-router-dom";

import { getFilm, updateFilm } from "@/utils/films";

import * as Style from "./style";

export async function loader({ params }) {
  const film = await getFilm(params.filmId);
  // Throw a 404 response in the loader
  if (!film) {
    throw new Response("", {
      status: 404,
      statusText: "Not Found",
    });
  }
  return film;
}

export async function action({ request, params }) {
  const formData = await request.formData();
  return updateFilm(params.filmId, {
    favorite: formData.get("favorite") === "true" ? "true" : "false",
  });
}

const Favorite = ({ film }) => {
  const fetcher = useFetcher();

  let { favorite } = film;
  // Optimistic UI:
  // If the fetcher has any formData being submitted, the star change to the new state immediately.
  // When the action is done, we're back to using the actual data.
  if (fetcher.formData) {
    favorite = fetcher.formData.get("favorite") === "true" ? "true" : "false";
  }

  return (
    <Style.FavoriteForm>
      <fetcher.Form method="post">
        <Style.FavoriteButton
          type="submit"
          name="favorite"
          value={favorite === "true" ? "false" : "true"}
          aria-label={
            favorite === "true" ? "Remove from favorites" : "Add to favorites"
          }
        >
          {favorite === "true" ? "★" : "☆"}
        </Style.FavoriteButton>
      </fetcher.Form>
    </Style.FavoriteForm>
  );
};

const Film = () => {
  const film = useLoaderData();
  const minutes = film.running_time % 60;
  const hours = Math.floor(film.running_time / 60);

  return (
    <Style.FilmContainer>
      <Style.FilmBanner src={film.movie_banner} alt="film banner" />
      <Style.Row style={{ justifyContent: "space-between" }}>
        <Style.Column style={{ width: "100%" }}>
          <Style.Row style={{ gap: "8px" }}>
            <Favorite film={film} />
            <Style.Title>{film.title}</Style.Title>
          </Style.Row>
          <Style.Row
            style={{
              width: "100%",
              gap: "32px",
            }}
          >
            <p>{film.release_date}</p>
            <Style.Score>{film.rt_score}</Style.Score>
            <p>{`${hours}h ${minutes}min`}</p>
          </Style.Row>
        </Style.Column>
        <Style.Column>
          <p style={{ whiteSpace: "nowrap" }}>
            <span style={{ color: "#87CEEB" }}>Director:</span>
            {" "}
            {film.director}
          </p>
          <p style={{ whiteSpace: "nowrap" }}>
            <span style={{ color: "#87CEEB" }}>Producer:</span>
            {" "}
            {film.producer}
          </p>
        </Style.Column>
      </Style.Row>
      <Style.Description>{film.description}</Style.Description>
    </Style.FilmContainer>
  );
};

Favorite.propTypes = {
  film: PropTypes.shape({
    favorite: PropTypes.string,
  }),
};
Favorite.defaultProps = {
  film: {},
};

export default Film;
