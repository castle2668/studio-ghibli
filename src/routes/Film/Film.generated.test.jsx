import React from "react";
import renderer from "react-test-renderer";

import Film, { action, loader } from "./Film";

jest.mock("react-router-dom");
jest.mock("@/utils/films");
jest.mock("./style");

const renderTree = (tree) => renderer.create(tree);

describe("<Film>", () => {
  it("should render component", async () => {
    expect(renderTree(<Film />).toJSON()).toMatchSnapshot();
  });
});

describe("loader", () => {
  it("should expose a function", () => {
    expect(loader).toBeDefined();
  });

  it("loader should return expected output", async () => {
    try {
      const res = await loader({});
      expect(res).toBeDefined();
    } catch (error) {
      expect(error).toBeDefined();
    }
  });
});

describe("action", () => {
  it("should expose a function", () => {
    expect(action).toBeDefined();
  });

  it("action should return expected output", async () => {
    const retValue = await action({});
    expect(retValue).toBeFalsy();
  });
});
