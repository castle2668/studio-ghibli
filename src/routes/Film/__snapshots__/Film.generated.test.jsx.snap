// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Film> should render component 1`] = `null`;
